docs/source-2.0/guides/client-guidance/application-protocols/http.md

```diff
@@ -0,0 +1,179 @@
+     * 
+     * Smithy clients know whether a given header should have a single value or
+     * a list value. This helper method simplifies usage for headers with a
+     * single value.
+     *
+     * @param name Name of the field to get.
+     * @return the matching field value, or null if not found.
+     */
+    default String getFirstValue(String name) {
+        var list = getAllValues(name);
+        return list.isEmpty() ? null : list.get(0);
+    }
+
+    /**
+     * Get the values of a specific field by case-insensitive name.
+     *
+     * @param name Name of the field to get the values of.
+     * @return the values of the field, or an empty list.
+     */
+    List<String> getAllValues(String name);
+}
+```
+
+#### Implementation recommendations
+
+It is not recommended to automatically attempt to join values for a given field
+key at the HTTP layer. {rfc}`9110#section-5` allows field values to be joined
+with a comma, but doing so automatically can introduce data corruption if one of
+the field values already includes a comma. {rfc}`Section 5.6 <9110#section-5.6>`
+includes productions that can help to handle those edge cases, but whether they
+are used or not is up to the protocol definition.
+
+### Requests and Responses
+
+{rfc}`9110` discusses HTTP requests and responses collectively as "messages",
+and it can be useful to encode their commonalities in a shared interface.
+
+```java
+public interface HttpMessage {
+    /**
+     * Get the headers of the message.
+     *
+     * @return headers.
+     */
+    HttpFields headers();
+
+    /**
+     * Get the body of the message, or null.
+     *
+     * @return the message body or null.
+     */
+    DataStream body();
+}
+```
+
+Requests introduce the method and URI properties.
+
+```java
+public interface HttpRequest extends HttpMessage {
+    /**
+     * Get the method of the request.
+     *
+     * @return the method.
+     */
+    String method();
+
+    /**
+     * Get the URI of the request.
+     *
+     * @return the request URI.
+     */
+    URI uri();
+}
+```
+
+Responses introduce a status code.
+
+```java
+public interface HttpResponse extends HttpMessage {
+    /**
+     * Get the status code of the response.
+     *
+     * @return the status code.
+     */
+    int statusCode();
+}
+```
+
+### Clients
+
+It is recommended to make HTTP clients implementations of
+[`ClientTransport`](#transport-clients).
+
+```java
+public interface HttpClient implements ClientTransport<HttpRequest, HttpResponse> {
+    HttpResponse send(Context context, HttpRequest request);
+}
+```
+
+#### Context
+
+HTTP clients don't have many common context parameters, but they should check
+the context for a request timeout setting and use it if it's present.
+
+```java
+public final class HttpContext {
+    public static final Context.Key<Duration> HTTP_REQUEST_TIMEOUT = Context.key("HTTP.RequestTimeout");
+
+    private HttpContext() {}
+}
+```
```

Comment:
In general, how do we feel about access modifiers? I recognize that good library design requires careful attention to which declarations are `public`, `internal`, `private`, etc. but these interfaces are intended to be illustrative, not necessarily copy-paste solutions. I lean towards omitting any unnecessary syntax which isn't useful in conveying the intent of the abstraction, including removing `public` from basically everything.